keras_nlp/models/whisper/whisper_backbone.py

```diff
@@ -0,0 +1,262 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Whisper backbone model."""
+
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.layers.position_embedding import PositionEmbedding
+from keras_nlp.layers.token_and_position_embedding import (
+    TokenAndPositionEmbedding,
+)
+from keras_nlp.models.backbone import Backbone
+from keras_nlp.models.whisper.whisper_decoder import WhisperDecoder
+from keras_nlp.models.whisper.whisper_encoder import WhisperEncoder
+
+
+def whisper_kernel_initializer(stddev=0.02):
+    return keras.initializers.TruncatedNormal(stddev=stddev)
+
+
+class WhisperBackbone(Backbone):
+    """Whisper encoder-decoder network for speech.
+
+    This class implements a Transformer-based encoder-decoder model as
+    described in
+    ["Robust Speech Recognition via Large-Scale Weak Supervision"](https://arxiv.org/abs/2212.04356).
+
+    The default constructor gives a fully customizable, randomly initialized Whisper
+    model with any number of layers, heads, and embedding dimensions. To load
+    preset architectures and weights, use the `from_preset` constructor.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/openai/whisper).
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer encoder layers and
+            transformer decoder layers.
+        num_heads: int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        hidden_dim: int. The size of the transformer encoding and pooler layers.
+        intermediate_dim: int. The output dimension of the first Dense layer in
+            a two-layer feedforward network for each transformer.
+        num_mels: int. The number of mel-frequency filters. For now, only 80
```

Comment:
This arg is a little confusing to me. Why only 80? And if there is only one option, are we better off hard coding this for now, and adding configurability where we are ready?

Adding an arg is never backwards incompatible.

Reply:
Not sure why only 80 is supported. The official repo mentions it several times in functions which load/process audio:

https://github.com/openai/whisper/blob/main/whisper/audio.py#L95
https://github.com/openai/whisper/blob/main/whisper/audio.py#L114

Ideally, any `n_mels` value should work **in the model**: https://github.com/openai/whisper/blob/main/whisper/model.py#L159. `n_mels` is essentially the number of input channels in the first conv layer: https://github.com/openai/whisper/blob/main/whisper/model.py#L148. So, this value could be anything.

Maybe, the restriction is a "loading audio" specific restriction and not a model restriction.

Reply:
I would remove this as an option until we are ready to support multiple value of it. Can always add later.